haystack/pipelines/base.py

```diff
@@ -1053,41 +1141,96 @@ def eval(
-                    sas, _ = semantic_answer_similarity(
+                    sas, _, pred_label_sas_grid = semantic_answer_similarity(
                         predictions=predictions,
                         gold_labels=gold_labels,
                         sas_model_name_or_path=sas_model_name_or_path,
                         batch_size=sas_batch_size,
                         use_gpu=sas_use_gpu,
                     )
                     df["sas"] = sas
+                    df["gold_answers_sas"] = [
+                        gold_answers_sas_per_pred[0] for gold_answers_sas_per_pred in pred_label_sas_grid
+                    ]
+                    df.map_rows = partial(df.apply, axis=1)
+                    df["sas_context_scope"] = df.map_rows(
+                        lambda row: max(
+                            sas
+                            for sas, sim in zip(
+                                row["gold_answers_sas"] + [0.0], row["gold_contexts_similarity"] + [100]
+                            )
+                            if sim > context_matching_threshold
+                        )
+                    )
+                    df["sas_document_scope"] = df.map_rows(
+                        lambda row: max(
+                            sas
+                            for sas, doc_match in zip(
+                                row["gold_answers_sas"] + [0.0], row["gold_documents_id_match"] + [1.0]
+                            )
+                            if doc_match == 1.0
+                        )
+                    )
+                    df["sas_document_and_context_scope"] = df.map_rows(
+                        lambda row: max(
+                            sas
+                            for sas, sim, doc_match in zip(
+                                row["gold_answers_sas"] + [0.0],
+                                row["gold_contexts_similarity"] + [100],
+                                row["gold_documents_id_match"] + [1.0],
+                            )
+                            if sim > context_matching_threshold and doc_match == 1.0
+                        )
+                    )
 
         # reorder columns for better qualitative evaluation
         for key, df in eval_result.node_results.items():
             desired_col_order = [
-                "multilabel_id",
+                "multilabel_id",  # generic
                 "query",
-                "filters",  # generic
-                "gold_answers",
+                "filters",
+                "gold_answers",  # answer-specific
                 "answer",
-                "context",
-                "exact_match",
+                "context",  # generic
+                "exact_match",  # answer-specific
                 "f1",
-                "sas",  # answer-specific
-                "gold_document_contents",
-                "content",
-                "gold_id_match",
+                "sas",
```

Comment:
isn't sas still answer-specific? What's the default meaning when there is neither `# answer-specific` nor `# generic`?

Reply:
Yes it is. We once had this list with multiple values per line, each line representing a scope and having the appropriate comment. With black that broke and I tried to mark the beginning of such a scope with the comment. So yes, anything after (and including) exact_match is answer-specific until gold_contexts. Do you think it's worth to set the comment to each line?

Reply:
done.